RFC: TPLG: Add topology parser to firmware

singalsu opened the pull request:
This request-for-comments only patch adds to Sue Creek platform firmware
a topology data parser that creates static pipelines those are defined
by the embedded topology file data.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

singalsu commented:
@lgirdwood @lyakh This is RFC PR is for review only. I have not been able to test it yet.

ranj063 commented on src/lib/topology-priv.h:
```diff
@@ -0,0 +1,301 @@
+/*
```
@singalsu do we need to duplicate this header file for the testbench and the firmware?

singalsu replied:
I was thinking to eventually remove the parser from test bench and use this instead. But keep in test bench e.g. the library load code and other specific for it that this parser would call when compiled for test bench.

ranj063 commented on src/lib/topology.c:
```diff
@@ -0,0 +1,1049 @@
+/*
```
@singalsu same for the parser too. The major difference between the two is one takes the the topology file as input and the other uses the bytes.

singalsu replied:
@ranj063 Yes, I'd prefer this way to keep the data in memory for both cases since it's much simpler. The fseek() pointer arithmetic is really tricky and needs lot of local copied parts.

lgirdwood replied:
@singalsu would it be simpler for your parser (and it would be smaller) if the parser was only parsing a file of IPC structures instead of the ALSA structure & tokens ? i.e.

 1) take the ALSA binary file strip out ALSA struct and convert to IPC structs
 2) load the new file with IPC structs into SOF.
 
Step 1 above is mostly done already in your patch and in the driver (topology.c) and means creating a small cmd line tool based on topology.c that does this conversion in userspace. This way your FW parser is far simpler as it will just call handler.c calls directly and pass in each structure ?

lgirdwood commented on src/lib/topology.c:
```diff
@@ -0,0 +1,1049 @@
+	}
+}
+
+void sof_parse_string_tokens(void *object,
```
can we prefix all non static functions tplg_ instead of sof_

singalsu replied:
OK, also there's a lot of function those can be static.

lgirdwood commented on src/tasks/audio.c:
```diff
@@ -52,8 +53,422 @@ struct audio_data {
+
+/* Temporarily place the topology into a C array */
+
+unsigned char test_tone_playback_tplg[] = {
```
This ptr will eventually come from linker script variable, but this is fine atm.

lgirdwood commented on src/tasks/audio.c:
```diff
@@ -52,8 +53,422 @@ struct audio_data {
+
 int do_task_master_core(struct sof *sof)
 {
+#if defined(CONFIG_SUECREEK)
```
Don't do an ifdef here, just check if variable != NULL (set by linker script), this meas we can easily add a payload to any FW.

lgirdwood commented on src/lib/topology.c:
```diff
@@ -0,0 +1,1049 @@
+#define DEBUG_PRINT(...) fprintf(stderr, __VA_ARGS__)
+#else
+#define DEBUG_PRINT(...) do { } while (0)
+#endif
```
Yep, new trace can do printf on host now :)

lgirdwood commented:
@singalsu has there been any updates here for static pipelines ?

lgirdwood commented:
A hostless pipeline should probably have some tokens to define this. i.e. some may start automatically with static rate/format, etc whilst other may be started via a kcontrol (like a switch) again using a static rate, format etc.

singalsu commented:
@lgirdwood I have updated the patch now. It's still the basic functionality with comments addressed, hope I got your ideas right. This is potentially possible to merge since the feature is opt-out. When not enabled there's a minimal topology parse function that only returns -EINVAL.

The tokens for start control sound like a good idea. @ranj063 has work done for controlling hostless tone pipeline that could be utilized.

michalgrodzicki commented:
UT failed,
please merge with the code from the top of master (UT are fied there)

singalsu commented:
OK, I will update this next week (ww51).

singalsu commented:
@lgirdwood Thanks for review! I'll update the PR after holidays in 1st week in Jan.

singalsu commented:
@plbossart This work was intended for Sue Creek to replace the not maintained static pipelines (the included topology is a tone pipeline to DAI for wakeup tests).

@lyakh @lgirdwood I'm myself wondering if it becomes redundant now that IPC over SPI is becoming functional. Then the normal topology download could work, right? In this case this could as easier-to-maintain version (no fseeks, less ASoC headers dependency) replace the topology parser in testbench.

lgirdwood commented:
@singalsu not redundant. We need method to statically define pipelines using current userspace tools and not just hacked into static_pipeline.c

singalsu commented:
@lgirdwood Thanks for confirm! I'll work next with this PR update.

lgirdwood commented:
@singalsu great - btw, pls remove existing static_pipeline.c code as part of your new PR

singalsu commented on src/lib/topology-priv.h:
```diff
@@ -0,0 +1,314 @@
+	uint32_t payload_size;
+	uint32_t index;
+	uint32_t count;
+};
```
OK, done!

singalsu commented on src/lib/topology.c:
```diff
@@ -0,0 +1,1208 @@
+
+			/* matched - now load token */
+			tokens[j].get_token(elem, object, tokens[j].offset,
+					    tokens[j].size);
```
Yep, done.

singalsu commented:
@lgirdwood I have just updated the PR. I've addressed your two comments from the second review round plus I fixed issues that I got when re-testing. The code loads the embedded topology now ok. The verbose trace looks like this (I can't yet print strings in xtensa version, in host/TB version with other gcc version it works). 

```
 CORE  LEVEL      COMP_ID                TIMESTAMP            DELTA                FILE_NAME	CONTENT
    0      1         PIPE             38714.322917     38714.324219           topology.c:1137	tplg_parse(), type = 8, size = 112, count = 1, index = 0
    0      1         PIPE             38720.312500         5.989583           topology.c:1187	tplg_parse(), MANIFEST
    0      1         PIPE             38723.802083         3.489583           topology.c:1137	tplg_parse(), type = 5, size = 2196, count = 7, index = 5
    0      1         PIPE             38729.375000         5.572917           topology.c:1144	tplg_parse(), DAPM_WIDGET, count = 7
    0      1         PIPE             38737.552083         8.177083           topology.c:1076	load_widget(), SIGGEN (0)
    0      1         PIPE             38743.125000         5.572917           topology.c:368 	parse_word_tokens(), token = 800, value = 48000
    0      1         PIPE             38749.635417         6.510417           topology.c:337 	parse_string_tokens(), token = 402
    0      1         PIPE             38769.114583        19.479166           topology.c:1102	load_widget(), num_kcontrols = 1
    0      1         PIPE             38772.708333         3.593750           topology.c:613 	load_controls()
    0      1         PIPE             38776.197917         3.489583           topology.c:628 	load_controls(), mixer_ctl
    0      1         PIPE             38781.718750         5.520833           topology.c:1033	load_widget(), PGA (1)
    0      1         PIPE             38787.968750         6.250000           topology.c:368 	parse_word_tokens(), token = 400, value = 2
    0      1         PIPE             38792.656250         4.687500           topology.c:368 	parse_word_tokens(), token = 401, value = 2
    0      1         PIPE             38797.291667         4.635417           topology.c:368 	parse_word_tokens(), token = 403, value = 0
    0      1         PIPE             38801.875000         4.583333           topology.c:337 	parse_string_tokens(), token = 402
    0      1         PIPE             38816.458333        14.583333           topology.c:1102	load_widget(), num_kcontrols = 1
    0      1         PIPE             38820.052083         3.593750           topology.c:613 	load_controls()
    0      1         PIPE             38823.489583         3.437500           topology.c:628 	load_controls(), mixer_ctl
    0      1         PIPE             38828.854167         5.364583           topology.c:1025	load_widget(), BUFFER (2)
    0      1         PIPE             38833.125000         4.270833           topology.c:368 	parse_word_tokens(), token = 100, value = 768
    0      1         PIPE             38837.604167         4.479167           topology.c:368 	parse_word_tokens(), token = 101, value = 65
    0      1         PIPE             38857.343750        19.739584           topology.c:1025	load_widget(), BUFFER (3)
    0      1         PIPE             38861.458333         4.114583           topology.c:368 	parse_word_tokens(), token = 100, value = 768
    0      1         PIPE             38865.989583         4.531250           topology.c:368 	parse_word_tokens(), token = 101, value = 65
    0      1         PIPE             38885.208333        19.218750           topology.c:1016	load_widget(), SCHEDULER (4)
    0      1         PIPE             38889.322917         4.114583           topology.c:368 	parse_word_tokens(), token = 200, value = 1000
    0      1         PIPE             38894.270833         4.947917           topology.c:368 	parse_word_tokens(), token = 201, value = 0
    0      1         PIPE             38899.270833         5.000000           topology.c:368 	parse_word_tokens(), token = 203, value = 0
    0      1         PIPE             38904.062500         4.791667           topology.c:368 	parse_word_tokens(), token = 204, value = 48
    0      1         PIPE             38908.906250         4.843750           topology.c:368 	parse_word_tokens(), token = 205, value = 0
    0      1         PIPE             38913.697917         4.791667           topology.c:368 	parse_word_tokens(), token = 202, value = 200000
    0      1         PIPE             38928.020833        14.322917           topology.c:1060	load_widget(), DAI (5)
    0      1         PIPE             38932.812500         4.791667           topology.c:368 	parse_word_tokens(), token = 155, value = 5
    0      1         PIPE             38938.906250         6.093750           topology.c:337 	parse_string_tokens(), token = 154
    0      1         PIPE             38944.583333         5.677083           topology.c:368 	parse_word_tokens(), token = 400, value = 0
    0      1         PIPE             38949.114583         4.531250           topology.c:368 	parse_word_tokens(), token = 401, value = 2
    0      1         PIPE             38953.750000         4.635417           topology.c:368 	parse_word_tokens(), token = 403, value = 2
    0      1         PIPE             38958.802083         5.052083           topology.c:337 	parse_string_tokens(), token = 402
    0      1         PIPE             39014.791667        55.989582           topology.c:1016	load_widget(), SCHEDULER (6)
    0      1         PIPE             39018.802083         4.010417           topology.c:368 	parse_word_tokens(), token = 200, value = 1000
    0      1         PIPE             39023.541667         4.739583           topology.c:368 	parse_word_tokens(), token = 201, value = 2
    0      1         PIPE             39028.489583         4.947917           topology.c:368 	parse_word_tokens(), token = 203, value = 0
    0      1         PIPE             39033.281250         4.791667           topology.c:368 	parse_word_tokens(), token = 204, value = 48
    0      1         PIPE             39038.125000         4.843750           topology.c:368 	parse_word_tokens(), token = 205, value = 0
    0      1         PIPE             39042.916667         4.791667           topology.c:368 	parse_word_tokens(), token = 202, value = 5000
    0      1         PIPE             39055.781250        12.864583           topology.c:1137	tplg_parse(), type = 10, size = 1776, count = 1, index = 0
    0      1         PIPE             39061.093750         5.312500           topology.c:1195	tplg_parse(), BACKEND_LINK
    0      1         PIPE             39064.583333         3.489583           topology.c:1137	tplg_parse(), type = 4, size = 528, count = 4, index = 5
    0      1         PIPE             39069.635417         5.052083           topology.c:1158	tplg_parse(), DAPM_GRAPH
    0      1         PIPE             39075.364583         5.729167           topology.c:462 	Connect 0 -> 2
    0      1         PIPE             39085.416667        10.052083           topology.c:462 	Connect 2 -> 1
    0      1         PIPE             39095.364583         9.947917           topology.c:462 	Connect 1 -> 3
    0      1         PIPE             39105.468750        10.104167           topology.c:462 	Connect 3 -> 5
    0      1         PIPE             39143.177083        37.708332           topology.c:1233	tplg_parse(), Done.

```

singalsu commented:
@lgirdwood @plbossart I didn't yet increase ABI minor version. There's tone generator tokens added. If it's needed I will need to make for Linux side PR too. Or remove those tokens from this to not violate sync. What do you think is the best choice now?

lgirdwood commented on src/include/sof/topology.h:
```diff
@@ -0,0 +1,39 @@
+#ifndef TOPOLOGY_H
+#define TOPOLOGY_H
+
+int tplg_parse(struct ipc *ipc, uint8_t *tplg_bytes, size_t tplg_size);
```
tplg_bytes should be a void *

lgirdwood commented on src/include/uapi/user/tokens.h:
```diff
@@ -111,6 +111,9 @@
 #define SOF_TKN_INTEL_DMIC_PDM_SKEW		706
 
+/* Tone */
+#define SOF_TKN_TONE_SAMPLE_RATE		800
```
is this missing from kernel headers ? If so we need to add it as a separate PR ASAP @plbossart fyi.

singalsu commented:
I'll close this an reopen later when the new approach is ready.
